### PR TITLE
[PORT]Throw exception when circular reference is detected in LG imports

### LINKED ISSUE #2340: [botbuilder-lg] [bug] Infinite Loop in Imports
## Versions
What package version of the SDK are you using. botbuilder-lg: 4.9.2
What nodejs version are you using v10.18.1
What browser version are you using
What os are you using WSL

## Describe the bug
Having two language generation files that import each other causes an infinite loop/exception.
It's not really the loop that concerns me, I assume this means that LG is reloading templates if I import a file twice throughout my structure so I'm concerned about the impact to loading time.

## To Reproduce
In this example, I get an exception but I can't even see the whole trace because it's so long. With my templates that I actually use where the circular imports might not be so direct, `parseFile` just hangs.
a.lg:
```Markdown
> Import b.lg
[b](./b.lg)
# a
- a
```

b.lg:
```Markdown
> Import a.lg
[a](./a.lg)
# b
- b
```

code.ts:
```TypeScript
import { Templates } from 'botbuilder-lg'
Templates.parseFile('a.lg')
```

## Expected behavior
Should not cause an infinite loop/exception.

[bug]

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #2340
Original PR:https://github.com/microsoft/botbuilder-dotnet/pull/4779

## Description
Having two language generation files that import each other causes an infinite loop.

## Specific Changes
Throw exception when occurs the cycle reference in LG import.
For example: there are two Lg files:
a.lg
```
> Import b.lg
[b](./b.lg)
# a
- a
```
and b.lg
```
> Import a.lg
[a](./a.lg)
# b
- b
```

When parsing the a.lg, using `Templates.ParseFile("a.lg")`, would get a "loop detect exception."

### CODE REVIEW

File: libraries/botbuilder-lg/src/templatesParser.ts [hunk 1]
```diff
@@ -151,18 +167,21 @@ export class TemplatesParser {
...
         expressionParser?: ExpressionParser,
-        cachedTemplates?: Map<string, Templates>
+        cachedTemplates?: Map<string, Templates>,
+        parentTemplates?: Templates[]
     ): Templates {
         if (!resource) {
             throw new Error('lg resource is empty.');
         }
         cachedTemplates = cachedTemplates || new Map<string, Templates>();
+        parentTemplates = parentTemplates || [];
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/templatesParser.ts:184]: Both of these can be expressed as defaults inside the parameter list.
```typescript
private static innerParseResource(..., cachedTemplates? = new Map<string, Templates>, parentTemplates?: Templates[] = [])
```

File: libraries/botbuilder-lg/src/templatesParser.ts [hunk 2]
```diff
@@ -215,9 +234,9 @@ export class TemplatesParser {
         return parser.file();
     }
 
-    private static getReferences(file: Templates, cachedTemplates?: Map<string, Templates>): Templates[] {
+    private static getReferences(file: Templates, cachedTemplates?: Map<string, Templates>, parentTemplates?: Templates[]): Templates[] {
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/templatesParser.ts:237]: Same suggestion to push defaults into the parameter list.

**Review by joshgummersall — COMMENTED** [review]
I left a few inline suggestions. Otherwise, this looks good!

File: libraries/botbuilder-lg/src/templatesParser.ts (hunk 1 above)
Comment (Danieladu) [reply libraries/botbuilder-lg/src/templatesParser.ts:184]: Really learn a lot from you @joshgummersall . Big thanks!

**Review by joshgummersall — APPROVED** [review]
Nice! Looks good to me.

**Pull request merged**